speculative_decoding/model/model.py

```diff
@@ -0,0 +1,142 @@
+class Model:
+    def __init__(self, **kwargs):
+        self._data_dir = kwargs["data_dir"]
+        self._config = kwargs["config"]
+        self._secrets = kwargs["secrets"]
+        self._request_id_counter = count(start=1)
+        self._triton_server = None
+        self._triton_client = None
+        if "openai-compatible" in self._config["model_metadata"]["tags"]:
+            raise NotImplementedError("openai-compatible")
+
+    def load(self):
+        tensor_parallel_count = self._config["model_metadata"].get(
+            "tensor_parallelism", 1
+        )
+        pipeline_parallel_count = self._config["model_metadata"].get(
+            "pipeline_parallelism", 1
+        )
+
+        env = {}
+        try:  # Secrets does not have __contains__.
+            hf_access_token = self._secrets["hf_access_token"]
+            env["HUGGING_FACE_HUB_TOKEN"] = hf_access_token
+        except Exception:  # `SecretNotFound` from truss template cannot be referenced.
+            hf_access_token = None
+
+        # Target model.
+        huggingface_hub.snapshot_download(
+            self._config["model_metadata"]["engine_repository"],
+            local_dir=os.path.join(TRITON_DIR, TARGET_MODEL_KEY, "1"),
+            local_dir_use_symlinks=False,
+            max_workers=4,
+            use_auth_token=hf_access_token,
+        )
+        # Draft model.
+        huggingface_hub.snapshot_download(
+            self._config["model_metadata"]["speculative_decoding"][
+                "draft_engine_repository"
+            ],
+            local_dir=os.path.join(TRITON_DIR, DRAFT_MODEL_KEY, "1"),
+            local_dir_use_symlinks=False,
+            max_workers=4,
+            use_auth_token=hf_access_token,
+        )
+
+        if not helpers.is_triton_server_alive():
+            self._triton_server = helpers.TritonServer(
+                TRITON_DIR,
+                parallel_count=tensor_parallel_count * pipeline_parallel_count,
+            )
+            self._triton_server.load_server_and_model(env)
+
+        client = triton_grpc.InferenceServerClient("0.0.0.0:8001")
+        self._target_model = spec_dec.ModelWrapper(
+            client,
+            TARGET_MODEL_KEY,
+            transformers.AutoTokenizer.from_pretrained(
+                self._config["model_metadata"]["tokenizer_repository"]
+            ),
+        )
+
+        self._draft_model = spec_dec.ModelWrapper(
+            client,
+            DRAFT_MODEL_KEY,
+            transformers.AutoTokenizer.from_pretrained(
+                self._config["model_metadata"]["speculative_decoding"][
+                    "draft_tokenizer_repository"
+                ]
+            ),
+        )
+
+    async def predict(self, model_input) -> str | AsyncGenerator[str, None]:
+        request_id = str(os.getpid()) + str(next(self._request_id_counter))
+        request = helpers.GenerationRequest.parse_obj(model_input)
+        max_num_draft_tokens: int = self._config["model_metadata"][
```

Comment:
Would be good to move these out from model_metadata and into the build key. Also worth using Pydantic for this so you don't need to string index.